File: apps/faithful-android/core/attendance/src/main/kotlin/io/faithform/faithful/attendance/QrScanning.kt
package io.faithform.faithful.attendance

import com.google.zxing.BinaryBitmap
import com.google.zxing.DecodeHintType
import com.google.zxing.NotFoundException
import com.google.zxing.PlanarYUVLuminanceSource
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.qrcode.QRCodeReader
import java.security.SecureRandom

/**
 * Scanning a check-in code, with every decision out of the camera's way.
 *
 * ## Why the decoder is here and not in `:app`
 *
 * `:core:attendance` is a pure-JVM module. Putting the QR decode here means it
 * runs under `gradlew :core:attendance:test` against a **real generated QR
 * image**, on an ordinary runner, with no emulator and no camera. The
 * alternative — an ML Kit or Play-services decoder in `:app` — would have been
 * a binary nothing in CI could exercise, and the honest write-up would have had
 * to say the decode was untested.
 *
 * What remains in `:app` is CameraX handing over a byte array. That is a thin
 * seam, and it is the only part of this feature the device runbook has to cover.
 */

// ---------------------------------------------------------------------------
// Permission
// ---------------------------------------------------------------------------

/**
 * The camera permission, as this feature needs to reason about it.
 *
 * Android's runtime permissions have a third state iOS does not: *denied, but
 * asking again would still show a prompt*. Collapsing that into "denied" is how
 * apps end up telling someone to open Settings when a tap would have done.
 */
enum class CameraPermissionState {
    /** **Never been asked.** Where the app must be until "Scan" is tapped. */
    NOT_REQUESTED,

    GRANTED,

    /** Denied once. Asking again still raises the system dialog. */
    DENIED_CAN_ASK_AGAIN,

    /**
     * Denied permanently, or blocked by policy. No dialog will appear, so the
     * only route is app settings.
     */
    DENIED_PERMANENTLY;

    val canScan: Boolean get() = this == GRANTED

    /** Whether asking would actually produce a dialog. */
    val promptWouldAppear: Boolean
        get() = this == NOT_REQUESTED || this == DENIED_CAN_ASK_AGAIN
}

/**
 * The camera, reduced to what scanning needs.
 *
 * **What is deliberately absent:** any member that returns an image, writes a
 * file, or touches the photo library. A QR scan needs a string, so the interface
 * cannot express saving a frame — a stronger guarantee than a rule saying not
 * to.
 */
interface QrScanningFacade {
    fun permissionState(): CameraPermissionState

    /**
     * Raises the camera dialog.
     *
     * **Only ever called from an explicit "Scan" action.** Never at launch,
     * during onboarding, while browsing a feed, or when enabling automatic
     * attendance.
     */
    suspend fun requestPermission(): CameraPermissionState

    /** Starts delivering decoded strings. Idempotent. */
    suspend fun start(onCode: (String) -> Unit)

    /** Stops analysis and releases the camera. */
    suspend fun stop()

    /** Whether this device has a usable camera at all. */
    fun isAvailable(): Boolean
}

// ---------------------------------------------------------------------------
// Decoding
// ---------------------------------------------------------------------------

/**
 * Turns one camera frame's luminance plane into a decoded string.
 *
 * The Y plane of a `YUV_420_888` image is exactly the greyscale bitmap a QR
 * decoder wants, so no colour conversion and no `Bitmap` allocation is needed —
 * which also means no image object is ever created that could be written
 * anywhere.
 *
 * **Nothing is retained.** The reader is reset between frames and the byte array
 * belongs to the caller; this object holds no reference to it after returning.
 */
class LuminanceQrDecoder {
    private val reader = QRCodeReader()
    private val hints = mapOf(DecodeHintType.TRY_HARDER to true)

    /**
     * @param plane the Y plane, row-major
     * @param rowStride bytes per row, which CameraX pads and is *not* always
     *   equal to the width — assuming it is produces a skewed image that
     *   silently never decodes
     */
    fun decode(
        plane: ByteArray,
        width: Int,
        height: Int,
        rowStride: Int = width,
    ): String? {
        if (width <= 0 || height <= 0) return null
        if (rowStride < width) return null
        if (plane.size < rowStride * height) return null

        val source = PlanarYUVLuminanceSource(
            plane, rowStride, height, 0, 0, width, height, false,
        )

        return try {
            reader.decode(BinaryBitmap(HybridBinarizer(source)), hints).text
        } catch (_: NotFoundException) {
            // The overwhelmingly common case: no code in this frame. Not an
            // error, and emphatically not something to log — a log line per
            // frame at thirty frames a second is a battery drain and a way for
            // a payload to reach a log by accident.
            null
        } catch (_: Exception) {
            // A checksum failure or a malformed symbol. Same treatment.
            null
        } finally {
            reader.reset()
        }
    }
}

// ---------------------------------------------------------------------------
// What a scanned string may be
// ---------------------------------------------------------------------------

sealed interface ScannedPayload {
    /** A Faithful check-in capability. **Opaque** to the client. */
    data class CheckInToken(val token: String) : ScannedPayload

    /** Something else: a Wi-Fi code, a URL, a business card. */
    data object Unrecognised : ScannedPayload
}

object ScannedPayloadReader {
    /**
     * The wire format's prefix. Matching it lets the scanner ignore the poster
     * beside the screen without a round trip, and **nothing more** — the token
     * is signed, and only the server can say whether it is real.
     */
    const val TOKEN_PREFIX = "FF1."

    /** Matches `MAX_TOKEN_LENGTH` on the server. */
    const val MAXIMUM_LENGTH = 1024

    fun read(raw: String): ScannedPayload {
        val trimmed = raw.trim()
        if (trimmed.length > MAXIMUM_LENGTH || !trimmed.startsWith(TOKEN_PREFIX)) {
            return ScannedPayload.Unrecognised
        }
        val parts = trimmed.split(".")
        if (parts.size != 4 || parts.any { it.isEmpty() }) return ScannedPayload.Unrecognised
        return ScannedPayload.CheckInToken(trimmed)
    }
}

// ---------------------------------------------------------------------------
// The typed fallback
// ---------------------------------------------------------------------------

/**
 * The short code, mirrored from the server.
 *
 * Duplicated deliberately rather than fetched: this shapes the keyboard and
 * stops a hopeless request before it is sent, and a client that had to ask the
 * server what a code looks like could do neither offline. The server normalises
 * and validates again; this is convenience, never authority.
 */
object ShortCodeEntry {
    const val ALPHABET = "BCDFGHJKLMNPQRTVWXY3479"
    const val LENGTH = 7

    /**
     * Folds case and drops separators. **Substitutes nothing** — every
     * character a substitution table would map is already absent from the
     * alphabet, so mapping one could only turn a typo into a different valid
     * code and check someone into a service they did not choose.
     */
    fun normalise(input: String): String =
        input.uppercase().filter { ALPHABET.contains(it) }.take(LENGTH)

    fun isComplete(input: String): Boolean = normalise(input).length == LENGTH
}

// ---------------------------------------------------------------------------
// Phases
// ---------------------------------------------------------------------------

sealed interface ScanPhase {
    /** Before "Scan" is tapped. **The camera is not running.** */
    data object Idle : ScanPhase
    data object RequestingPermission : ScanPhase
    data object Scanning : ScanPhase
    /** A code is with the server. The camera is already stopped. */
    data object Submitting : ScanPhase
    data class Finished(val outcome: ScanOutcome) : ScanPhase
    data class Blocked(val block: ScanBlock) : ScanPhase
}

sealed interface ScanOutcome {
    val message: String

    data class Counted(override val message: String) : ScanOutcome
    data class AlreadyCounted(override val message: String) : ScanOutcome
    data class Refused(override val message: String) : ScanOutcome

    /**
     * **The only place a scan may be called a success.**
     *
     * Derived from the server's `outcome`, never from having read a code. A
     * phone that scanned a valid code and lost the network has checked nobody
     * in, and saying otherwise is a lie the person discovers when the church's
     * report disagrees with them.
     */
    val isSuccess: Boolean
        get() = this is Counted || this is AlreadyCounted
}

enum class ScanBlock {
    CAMERA_DENIED_CAN_ASK,
    CAMERA_DENIED_PERMANENTLY,
    CAMERA_UNAVAILABLE,
    OFFLINE,
}

/**
 * Whether a decoded string should be acted on.
 *
 * An image analyser runs on every frame — the same code decodes dozens of times
 * a second while it is in view, and a rotating display puts a *new* code up
 * every thirty seconds. Without this, one glance at a projector would spend a
 * person's whole attempt budget in under a second.
 */
data class ScanDebounce(
    private val lastCode: String? = null,
    private val lastAtEpochMillis: Long? = null,
) {
    fun shouldSubmit(code: String, nowEpochMillis: Long): Boolean {
        if (lastCode == null || lastAtEpochMillis == null || lastCode != code) return true
        return nowEpochMillis - lastAtEpochMillis >= REPEAT_WINDOW_MILLIS
    }

    fun recording(code: String, nowEpochMillis: Long): ScanDebounce =
        copy(lastCode = code, lastAtEpochMillis = nowEpochMillis)

    companion object {
        const val REPEAT_WINDOW_MILLIS = 10_000L
    }
}

/**
 * A fresh identity for one scan.
 *
 * **Random, every time.** Prompt 7 learned this the expensive way on the
 * geofence path: an idempotency key derived from stable inputs made a single
 * early refusal permanent for the rest of the service, because every subsequent
 * attempt replayed the refusal instead of being judged.
 */
object ScanAttemptIdentity {
    private val random = SecureRandom()

    fun make(): String {
        val bytes = ByteArray(16)
        random.nextBytes(bytes)
        return "scan-" + bytes.joinToString("") { "%02x".format(it) }
    }

    /** The attempt id *is* the key. A retry reuses it; a new tap makes a new one. */
    fun idempotencyKey(attemptId: String): String = "qr-$attemptId"
}

File: apps/faithful-ios/Sources/FaithfulKit/Attendance/QrScanning.swift
import Foundation

/// Scanning a check-in code, with every decision out of the camera's way.
///
/// The framework part of a scanner is small: open a session, receive strings,
/// stop. Everything that could be *wrong* is here instead, in plain Swift with
/// no `AVFoundation` import, so it can be exercised on any machine — the
/// permission sequence, what counts as a plausible payload, when a duplicate
/// read is the same code arriving twice, and what the person is told.
///
/// The camera adapter is `AVFoundationScanner`, and it makes no decisions.

// MARK: - Permission

/// The camera permission, as this feature needs to reason about it.
///
/// Faithful's own enum rather than `AVAuthorizationStatus`, for the reason
/// Prompt 7 established for Core Location: nothing above the adapter should
/// depend on a framework type, and the framework type is unconstructible on a
/// test runner.
public enum CameraAuthorization: String, Equatable, Sendable {
    /// **Never been asked.** This is the state the app must be in until the
    /// person taps "Scan". It is not an error and it is not a prompt.
    case notDetermined
    case denied
    case restricted
    case authorized

    public var canScan: Bool { self == .authorized }

    /// Whether asking would actually produce a prompt. iOS shows the camera
    /// alert exactly once; after that, only Settings can change the answer, and
    /// telling someone to "allow camera access" when no prompt will appear is
    /// worse than telling them nothing.
    public var promptWouldAppear: Bool { self == .notDetermined }
}

/// The camera, reduced to what scanning needs.
///
/// **What is deliberately absent:** any photo-library member, any capture-to-
/// disk member, and anything that returns an image. A QR scan needs a string
/// and nothing else, so the protocol cannot express saving a frame — which is
/// a stronger guarantee than a rule saying not to.
public protocol QrScanningFacade: Actor {
    func currentAuthorization() -> CameraAuthorization

    /// Raises the camera prompt.
    ///
    /// **Only ever called from an explicit "Scan" action.** Never at launch,
    /// during onboarding, while browsing a feed, or when enabling automatic
    /// attendance — `CheckInScanCoordinator` is the only caller and it is only
    /// reachable from that button.
    func requestAccess() async -> CameraAuthorization

    /// Starts delivering decoded strings. Idempotent.
    func start(onCode: @Sendable @escaping (String) -> Void) async throws

    /// Stops the session and releases the camera.
    func stop() async

    /// Whether this device has a usable camera at all. False in a simulator and
    /// on hardware with none — distinct from being denied, and a different
    /// message.
    func isAvailable() -> Bool
}

// MARK: - What a scanned string may be

/// What the scanner made of a decoded string.
public enum ScannedPayload: Equatable, Sendable {
    /// A Faithful check-in capability. **Opaque** — the client cannot read it,
    /// and the shape check below is a cheap filter, not a validation.
    case checkInToken(String)
    /// Something else entirely: a Wi-Fi code, a URL, a business card.
    case unrecognised
}

public enum ScannedPayloadReader {
    /// The wire format's prefix. Matching it lets the scanner ignore the poster
    /// beside the screen without a round trip, and *nothing more* — the token is
    /// signed, and only the server can say whether it is real.
    public static let tokenPrefix = "FF1."

    /// Matches `MAX_TOKEN_LENGTH` on the server. A longer string is refused
    /// there, so sending one would spend an attempt from the person's budget to
    /// learn what is already known here.
    public static let maximumLength = 1024

    public static func read(_ raw: String) -> ScannedPayload {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count <= maximumLength, trimmed.hasPrefix(tokenPrefix) else {
            return .unrecognised
        }
        // Four dot-separated parts, none empty. The signature is what actually
        // decides; this only avoids a pointless request.
        let parts = trimmed.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 4, !parts.contains(where: { $0.isEmpty }) else {
            return .unrecognised
        }
        return .checkInToken(trimmed)
    }
}

// MARK: - The typed fallback

/// The short code, mirrored from the server.
///
/// Duplicated deliberately rather than fetched: this is used to shape the
/// keyboard and to stop a hopeless request before it is sent, and a client that
/// had to ask the server what a code looks like could not do either offline.
/// The server normalises and validates again; this is convenience, never
/// authority. `tests/security/checkin-authority.test.ts` asserts the two
/// alphabets have not drifted apart.
public enum ShortCodeEntry {
    public static let alphabet = "BCDFGHJKLMNPQRTVWXY3479"
    public static let length = 7

    /// Folds case and drops separators. **Substitutes nothing** — every
    /// character a substitution table would map is already absent from the
    /// alphabet, so mapping one could only turn a typo into a different valid
    /// code and check someone into a service they did not choose.
    public static func normalise(_ input: String) -> String {
        String(input.uppercased().filter { alphabet.contains($0) }).prefix(length).description
    }

    public static func isComplete(_ input: String) -> Bool {
        normalise(input).count == length
    }
}

// MARK: - The scan session

/// What the person sees while scanning.
public enum ScanPhase: Equatable, Sendable {
    /// Before "Scan" is tapped. **The camera is not running and has not been
    /// asked for.**
    case idle
    /// The permission prompt is on screen.
    case requestingPermission
    case scanning
    /// A code was read and is with the server. The camera is already stopped.
    case submitting
    /// The server answered. `counted` is the only success.
    case finished(ScanOutcome)
    case blocked(ScanBlock)
}

public enum ScanOutcome: Equatable, Sendable {
    case counted(message: String)
    case alreadyCounted(message: String)
    case refused(message: String)

    /// **The only place a scan may be called a success.**
    ///
    /// Derived from the server's `outcome`, never from having read a code.
    /// A phone that scanned a perfectly valid code and lost the network has not
    /// checked anyone in, and saying otherwise would be a lie that a person
    /// only discovers when the church's report disagrees with them.
    public var isSuccess: Bool {
        switch self {
        case .counted, .alreadyCounted: return true
        case .refused: return false
        }
    }
}

public enum ScanBlock: String, Equatable, Sendable {
    case cameraDenied
    case cameraRestricted
    case cameraUnavailable
    case offline
}

/// Whether a decoded string should be acted on.
///
/// A metadata output fires continuously — the same code arrives many times a
/// second while it is in frame, and a rotating display puts a *new* code up
/// every thirty seconds. Without this, one glance at a projector would spend a
/// person's whole attempt budget in under a second.
public struct ScanDebounce: Equatable, Sendable {
    /// How long the same string is ignored after being acted on.
    public static let repeatWindow: TimeInterval = 10

    private var lastCode: String?
    private var lastAt: Date?

    public init() {}

    public func shouldSubmit(_ code: String, at now: Date) -> Bool {
        guard let lastCode, let lastAt, lastCode == code else { return true }
        return now.timeIntervalSince(lastAt) >= Self.repeatWindow
    }

    public func recording(_ code: String, at now: Date) -> ScanDebounce {
        var next = self
        next.lastCode = code
        next.lastAt = now
        return next
    }
}

/// A fresh identity for one scan.
///
/// **Random, every time.** Prompt 7 learned this the expensive way on the
/// geofence path: an idempotency key derived from stable inputs made a single
/// early refusal permanent for the rest of the service, because every
/// subsequent attempt replayed the refusal instead of being judged. A person who
/// scans, is refused, fixes the problem and scans again must get a *new*
/// verdict — so a new tap is a new identity.
public enum ScanAttemptIdentity {
    public static func make() -> String {
        // 16 random bytes, hex. Not derived from the token, the occurrence, or
        // the account — deriving from any of those is what caused the bug.
        var bytes = [UInt8](repeating: 0, count: 16)
        for index in bytes.indices { bytes[index] = UInt8.random(in: 0...255) }
        return "scan-" + bytes.map { String(format: "%02x", $0) }.joined()
    }

    public static func idempotencyKey(for attemptId: String) -> String {
        // The attempt id *is* the key. A retry of the same submission reuses it
        // and finds the first attempt; a new tap generates a new one.
        "qr-" + attemptId
    }
}
